include/onnxruntime/core/framework/sparse_tensor.h

```diff
@@ -1,72 +1,434 @@
+  /// if the value returned in kUndefined, the instance is not populated
+  /// </summary>
+  /// <returns>format enum</returns>
+  SparseFormat Format() const noexcept {
+    return format_;
+  }
+
+  /// <summary>
+  /// Returns a would be dense_shape
+  /// </summary>
+  /// <returns>reference to dense_shape</returns>
+  const TensorShape& DenseShape() const noexcept {
+    return dense_shape_;
+  }
+
+  /// <summary>
+  /// Calculates and returns how much this fully initialized SparseTensor data (would)
+  /// occupy in a contiguous allocation block, or, in fact, occupies if it owns the buffer.
+  /// </summary>
+  /// <returns>required allocation size</returns>
+  int64_t RequiredAllocationSize() const noexcept;
+
+  /// <summary>
+  /// Returns Tensor element type enum.
+  /// Useful for type dispatching
+  /// </summary>
+  /// <returns></returns>
+  int32_t GetElementType() const {
+    return ml_data_type_->GetDataType();
   }
 
-  const TensorShape& Shape() const {
-    return shape_;
+  /// <summary>
+  /// Return Element MLDataType
+  /// </summary>
+  /// <returns></returns>
+  MLDataType DataType() const noexcept {
+    return ml_data_type_;
   }
 
-  Tensor& MutableValues() {
-    return values_;
+  /// <summary>
+  /// Test for string type
+  /// </summary>
+  /// <returns>true if tensor values are strings</returns>
+  bool IsDataTypeString() const {
+    return utils::IsPrimitiveDataType<std::string>(ml_data_type_);
   }
 
-  Tensor& MutableIndices() {
-    return indices_;
+  /// <summary>
+  /// Checks if the Tensor contains data type T
+  /// </summary>
+  /// <typeparam name="T"></typeparam>
+  /// <returns>true if tensor contains data of type T</returns>
+  template <class T>
+  bool IsDataType() const {
+    return utils::IsPrimitiveDataType<T>(ml_data_type_);
   }
 
-  //TensorShape& MutableShape() {
-  //  return shape_;
-  //}
+  const OrtMemoryInfo& Location() const noexcept { return location_; }
+
+  /// <summary>
+  /// Read only access to Coo indices
+  /// </summary>
+  class CooView {
+   public:
+    explicit CooView(const Tensor& indices) noexcept
+        : indices_(indices) {}
+    const Tensor& Indices() const noexcept { return indices_; }
+
+   private:
+    std::reference_wrapper<const Tensor> indices_;
+  };
+
+  /// <summary>
+  /// Returns Coo index view
+  /// </summary>
+  /// <returns>CooView instance</returns>
+  CooView AsCoo() const;
+
+  /// <summary>
+  /// Uses COO index contained in the user allocated buffer along with the values buffer passed on
+  /// to the constructor. The buffer is used as is and its lifespan must eclipse the lifespan of the sparse
+  /// tensor instance. The OrtMemoryInfo (location) of the index is assumed to be the same as values.
+  ///
+  /// The index size must either exactly match the number of values in which case
+  /// index shape would be 1-D (values_count) or it must be twice the number of values
+  /// in which case its shape would be 2-D (values_count, 2)
```

Comment:
I guess that means in row-major order over (values_count, 2)?

Reply:
Yes, row major order everywhere